numpy/lib/stride_tricks.py

```diff
@@ -111,6 +111,133 @@ def as_strided(x, shape=None, strides=None, subok=False, writeable=True):
+        otherwise `shape[i]` will refer to axis `i` of `x`.
+
+    subok : bool, optional
+        If True, sub-classes will be passed-through, otherwise the returned
+        array will be forced to be a base-class array (default).
+
+    writeable : bool, optional
+        When true, allow writing to the returned view. The default is false,
+        as this should be used with caution: the returned view contains the
+        same memory location multiple times, so writing to one location will
+        cause others to change.
+
+    Returns
+    -------
+    view : ndarray
+        Sliding window view of the array.
+
+    See Also
+    --------
+    as_strided: Create a view into the array with the given shape and strides.
+    broadcast_to: broadcast an array to a given shape.
+
+    Notes
+    -----
+    For some cases there may be more efficient approaches to calculate
+    transformations across multi-dimensional arrays, for instance
+    `scipy.signal.fftconvolve`, where combining the iterating step with the
+    calculation itself while storing partial results can result in significant
+    speedups.
+
+    Examples
+    --------
+    >>> x = np.arange(5)
+    >>> sliding_window_view(x, 2)
+    array([[0, 1],
+           [1, 2],
+           [2, 3],
+           [3, 4]])
+
+    This also works in more dimensions, e.g.
+
+    >>> i, j = np.ogrid[:3, :4]
+    >>> x = 10*i + j
+    >>> x
+    array([[ 0,  1,  2,  3],
+           [10, 11, 12, 13],
+           [20, 21, 22, 23]])
+    >>> shape = (2,2)
+    >>> sliding_window_view(x, shape)
+    array([[[[ 0,  1],
+             [10, 11]],
+            [[ 1,  2],
+             [11, 12]],
+            [[ 2,  3],
+             [12, 13]]],
+           [[[10, 11],
+             [20, 21]],
+            [[11, 12],
+             [21, 22]],
+            [[12, 13],
+             [22, 23]]]])
+
+    The axis can be specified explicitly:
+
+    >>> sliding_window_view(x, 3, 0)
+    array([[[ 0, 10, 20],
+            [ 1, 11, 21],
+            [ 2, 12, 22],
+            [ 3, 13, 23]]])
+
+    The same axis can be used several times:
+
+    >>> sliding_window_view(x, (2, 3), (1, 1))
+    array([[[[ 0,  1,  2],
+             [ 1,  2,  3]]],
+           [[[10, 11, 12],
+             [11, 12, 13]]],
+           [[[20, 21, 22],
+             [21, 22, 23]]]])
+    """
+    shape = tuple(shape) if np.iterable(shape) else (shape,)
+    # first convert input to array, possibly keeping subclass
+    x = np.array(x, copy=False, subok=subok)
+
+    shape_array = np.array(shape)
+    if np.any(shape_array < 0):
+        raise ValueError('`shape` cannot contain negative value')
+    if np.any(x.shape < shape_array):
+        raise ValueError(
+            'window shape cannot be larger than input array shape')
+
+    if axis is None:
+        axis = np.arange(x.ndim)
+    axis = tuple(axis) if np.iterable(axis) else (axis,)
+
+    w_ndim = len(shape)
+    assert w_ndim == len(axis)
```

Comment:
```suggestion
    if len(shape) != len(axis):
        raise ValueError(f"Must provide matching length shape and axes; got {len(shape)} shape elements and {len(axes)} axes elements)
```

It might be nice to have a separate error for `axis=None`, to explain to the user that if `axis=None`, all the dimensions are expected.